sdk/provisioning/Azure.Provisioning/src/Expressions/BicepFunction.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using Azure.Core;
+using Azure.Provisioning.Resources;
+
+namespace Azure.Provisioning.Expressions;
+
+// At the moment I'm trying to keep this minimal and not pull everything in
+// https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/bicep-functions
+// but I expect it'll grow a fair bit over time.
+
+public static class BicepFunction
+{
+    public static BicepValue<string> GetUniqueString(params BicepValue<string>[] values) =>
+        BicepSyntax.Call("uniqueString", values.Select(v => v.Compile()).ToArray());
+
+    public static BicepValue<string> Take(BicepValue<string> text, BicepValue<int> size) =>
+        BicepSyntax.Call("take", text.Compile(), size.Compile());
+
+    public static BicepValue<string> CreateGuid(params BicepValue<string>[] values) =>
+        BicepSyntax.Call("guid", values.Select(v => v.Compile()).ToArray());
+
+    public static BicepValue<ResourceIdentifier> GetSubscriptionResourceId(params BicepValue<string>[] values) =>
+        BicepSyntax.Call("subscriptionResourceId", values.Select(v => v.Compile()).ToArray());
+
+    // TODO: Use the more efficient interpolated string handler
+    public static BicepValue<string> Interpolate(FormattableString text)
+    {
```

Comment:
We can also make use of Bicep functions via BicepValue to mix and match C# literals and Bicep expressions.  For example:

```C#
Resource parent = ...;
string prefix = "foo";
int maxLength = 20;
BicepOutput name = ...;
name = BicepFunction.Take(BicepFunction.Interpolate($"{prefix}-{BicepFunction.UniqueString(parent.Id)}", maxLength);
```

would compile into:

`name = take('foo-${uniqueString(parent.id)}', 20)`

with the C# values fully evaluated and Bicep references preserved.